LLVM Bitcode Linker: A self contained linker for nvptx and other targets

kjetilkjeka opened the pull request:
This PR introduces a new linker named `llvm-bitcode-linker`. It is a `self-contained` linker that can be used to link programs in `llbc` before optimizing and compiling to native code. It will first be used internally in the Rust compiler to enable tests for the `nvptx64-nvidia-cuda` target as the original `rust-ptx-linker` is deprecated. It will then be provided to users of the `nvptx64-nvidia-cuda` target with the purpose of linking ptx. More targets than nvptx will also be supported eventually.

The PR introduces a new unstable `LinkerFlavor` for the compiler. The compiler will also not be shipped with rustc but most likely instead be shipped in it's own unstable component (a follow up PR will be opened for this). This means that merging this PR should not add any stability guarantees.

When more details of `self-contained` is implemented it will only be possible to use the linker when `-Clink-self-contained=+linker` is passed.

<details>
  <summary>Original Description</summary>
  
**When this PR was created it was focused a bit differently. The original text is preserved here in case there's some interests in it**
  
I have experimenting with approaches to replace the ptx-linker and enable the nvptx target tests again. I think it's time to get some feedback on the approach.

### The problem
The only useful linker for the nvptx target is [this crate](https://github.com/denzp/rust-ptx-linker). Since this linker performs linking on llvm bitcode it needs to track the llvm version of rustc and use the same format. It has not been maintained for 3+ years and must be considered abandoned. Over the years rust have upgraded LLVM while the linker has been left to bitrot. It is no longer in a usable state.

Due to the difficulty of keeping the ptx-linker up to date outside of tree the nvptx tests was [disabled a long time ago](https://github.com/rust-lang/rust/commit/f8f9a2869cce570c994d96afb82f4162b1b44cca). It was [previously discussed](https://github.com/rust-lang/rust/pull/96842#issuecomment-1146470177) if adding the ptx-linker to the rust repo would be a possibility. My efforts in doing this stopped at getting an answered if the license would prohibit it from inclusion in the [Rust repo](https://github.com/rust-lang/rust/pull/96842#issuecomment-1148397554). I therefore concluded that a re-write would be necessary.

### The possible solution presented here
The llvm tools know perfectly well how to link and optimize llvm bitcode. Each of them only perform a single task, and are therefore a bit cumbersome to call with the current linker approach rustc takes. 

This PR adds a simple tool (current name `embedded-linker`) which can link self contained (often embedded) programs in llvm bitcode before compiling to the target format. Optimization will also be performed if lto is enabled. The rust compiler will make a single invocation to this tool, while the tool will orchestrate the many calls to the llvm tools.

### The questions
 - Is having control over the nvptx linking and therefore also tests worth it to add such tool? or should the tool live outside the rust repo?
 - Is the approach of calling llvm tools acceptable? Or would we want to keep the ptx-linker approach of using the llvm library? The tools seems to provide more simplicity and stability, but more intermediate files are being written. Perhaps there also are some performance penalty for the calling tools approach.
 - What is the process for adding such tool? MCP?
 - Does adding `llvm-link` to the llvm-tool component require any process?
 - Does it require some sort of FCP to remove ptx-linker as the default linker for ptx? Or is it sufficient that using the upstream ptx-linker is broken in its current state. it is possible to use a somewhat patched version of ptx-linker.
</details>

Commit:
LLVM_TOOLS: Include llvm-link as a llvm tool

kjetilkjeka commented:
@rustbot label +O-NVPTX

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-15`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6710309610/job/18235196369) [(plain)](https://github.com/rust-lang/rust/commit/5a9b1dc50fb22293872c7dc3f0ffeef8ca86c109/checks/18235196369/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=kjetilkjeka
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_af389c31-3683-49f4-ab37-5e8a345f5270
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=727964644349ee6a8b3a532d8207d0270fb5ff9d
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_af389c31-3683-49f4-ab37-5e8a345f5270
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_af389c31-3683-49f4-ab37-5e8a345f5270
GITHUB_TRIGGERING_ACTOR=kjetilkjeka
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/117458/merge
GITHUB_WORKFLOW_SHA=727964644349ee6a8b3a532d8207d0270fb5ff9d
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
........................................................

failures:

---- spec::tests::nvptx64_nvidia_cuda stdout ----
thread 'spec::tests::nvptx64_nvidia_cuda' panicked at compiler/rustc_target/src/spec/tests/tests_impl.rs:31:9:
  left: true
 right: false
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

```

</details>

oli-obk commented:
r? wg-codegen

oli-obk commented:
r? codegen

oli-obk commented:
cc @nagisa and @Zoxc tor previous involvement

kjetilkjeka commented:
Cc: @workingjubilee as I remember we talked about this in this issue https://github.com/denzp/rust-ptx-linker/issues/34#issuecomment-1658273214

bjorn3 commented:
Can we require fat LTO for nvptx? When rustc does fat LTO it already combines all crates into a single object file.

juntyr commented:
I'm really excited to hear that PTX might gain a maintained linker again!

Two years ago I was using Rust to write single-source kernels for NVIDIA GPUs, where LTO and inlining were crucial. I at some point had to fork the ptx-linker crate to keep LTO working with newer LLVM versions. It would be awesome if linking would "just work" instead again, so that the linker's is kept in sync and supports fat LTO.

Thank you so much for working on this ❤️

kjetilkjeka commented:
> Can we require fat LTO for nvptx? When rustc does fat LTO it already combines all crates into a single object file.

It would certainly be possible, rust is capable of calling the right functions in llvm to make it work. But it doesn't work "out of the box" and I'm not certain it's the correct thing to do long term eiteher.

If we decided to use only `lto="fat"`  we would disallow `lto="thin"` which helps a lot with compile times. One could argue that the general `ptx` program is small enough for this to be acceptable, but it doesn't feel right to leave such an optimization out without an excellent reason. Perhaps a bigger problem is that `linker-plugin-lto` would not work which would make cross-language lto more difficult.

When it comes to compiler change we would still need to add `llc` as a linker known to the compiler. This is required to compile from llvm-bitcode to native. The linker would be a unconventional one as it would not actually be able to link anything and only accept a single bitcode file as input.

When I tested it just now it seems like `libcompiler_builtins` still produces an .rlib which is linked by the linker even when `lto = "fat"`. I'm not sure if this is a bug or expected behavior, but it must be resolved before we may assume that the linker is only given a single bitcode file. (Note: `-Zbuild-std=core` is required due to not distributing this library and LTO towards libcore is required to work around some bugs right now and desired at the long term since calling functions on ptx is tremendously expensive.)

As an addition, there might be future benefits an approach as this embedded-linker can have to other targets than ptx as well. Maybe it can be used to bring thin lto to embedded targets where the native linker doesn't support it. It can also provide compilation without requiring to install an external target specific linker.

Is this an approach we would like to explore further even with the known deficiencies listed above? Is the main advantage that we avoid yet another tool that needs to be maintained? Or that we get more control using libLLVM compared to calling the llvm tools?

bjorn3 commented:
> When I tested it just now it seems like libcompiler_builtins still produces an .rlib which is linked by the linker even when lto = "fat".

That will be fixed soon. There is an open PR for making compiler-builtins participate in LTO.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-15`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6719990055/job/18262634081) [(plain)](https://github.com/rust-lang/rust/commit/282b8d38decf4e0d64e3d85c8ad0307d0f21b8fb/checks/18262634081/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=kjetilkjeka
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_1b31f66c-1327-4cd5-a70b-6ca4f8d69be3
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=66c06df0f67495dbb260c3e65030a44cb2d8136b
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_1b31f66c-1327-4cd5-a70b-6ca4f8d69be3
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_1b31f66c-1327-4cd5-a70b-6ca4f8d69be3
GITHUB_TRIGGERING_ACTOR=kjetilkjeka
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/117458/merge
GITHUB_WORKFLOW_SHA=66c06df0f67495dbb260c3e65030a44cb2d8136b
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
configure: rust.deny-warnings   := False
configure: profile              := dist
configure: build.configure-args := ['--set', 'rust.deny-warnings=false']
configure: rust.deny-warnings   := False
    build = serialize_and_parse(["--enable-full-tools"])
  File "/checkout/src/bootstrap/bootstrap_test.py", line 27, in serialize_and_parse
    section_order, sections, targets = configure.parse_args(configure_args)
  File "/checkout/src/bootstrap/configure.py", line 274, in parse_args
    return parse_example_config(known_args, config)
  File "/checkout/src/bootstrap/configure.py", line 431, in parse_example_config
    configure_file(sections, top_level_keys, targets, config)
  File "/checkout/src/bootstrap/configure.py", line 507, in configure_file
    configure_section(sections[section_key], section_config)
  File "/checkout/src/bootstrap/configure.py", line 483, in configure_section
    raise RuntimeError("failed to find config line for {}".format(key))

----------------------------------------------------------------------
Ran 12 tests in 0.057s

```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #117435) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

kjetilkjeka commented:
There are a few questions in my original post, but I think the one that stops me from moving forward is figuring out whether this is an approach that can be accepted. Are there any processes required to get something like this MR approved (MCP?) or is it just a matter of completing the implementation?

Insight like "this is probably not going to be accepted, but bjorn3's alternative probably will be" would also be very helpful to help me avoid putting effort into something that is probably not going to go anywhere.

petrochenkov commented:
I planned to review it in a couple of weeks.
Generally, if the new tool is compatible with the previous ptx-linker then it doesn't need a new linker flavor, it can reuse the old one.
Also, whether the tool is looked up inside the rustc toolchain (the new linker) or elsewhere (e.g. the old ptx-linker) is controlled by the `-C link-self-contained` option, you may check how it works as well.

petrochenkov commented:
I think the general approach is fine, we can ship a small additional tool, especially if it's not enabled by default.
The linker just needs to reuse the old PTX linker flavor and the linker's location should be determined through `-C link-self-contained`, as I said in the comment above.
@rustbot author

kjetilkjeka commented:
Thanks @petrochenkov!

There's a few practical problems with the current format of the old linker flavor. The most important is that the `link_rlib` and `link_whole_rlib` gives the same argument to the linker and is therefore impossible to separate from the linker. This difference needs to be used to figure out which libraries can be internalized and which cannot when running the linker. The old linker unfortunately internalizes everything except kernels which is a workaround but not always the desired behavior (e.g. not when having cpp kernels call rust device function or vice versa).

Does it then mean that I can change the current `PtxLinker` implementation to work like the current interface of the embedded linker introduced in this PR? Relevant context is that the old `ptx-linker` is currently broken on rust, but it is possible to use it with a few workaround. It's difficult to say exactly how many users relies on patched `ptx-linkers` on recent nightly compilers. I'm quite confident it's very very few, but I expect there to be some.

petrochenkov commented:
>Relevant context is that the old `ptx-linker` is currently broken when used from rust, but it is possible to use it in a degraded state with a relatively simple patch. It's difficult to say exactly how many users relies on patched `ptx-linkers` on recent nightly compilers. I'm quite confident it's very very few, but I expect there to be some.

If the old ptx linker already requires patches to get into a working state, then a wrapper script mapping `--whole-rlib` into `--rlib` also won't be a problem.
If that project is revived and receives updates, then there will be an opportunity for interface convergence.
So I suggest to just make the changes to `PtxLinker` when strictly necessary, but keep compatibility when possible.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7036389536/job/19148866328) [(plain)](https://github.com/rust-lang/rust/commit/2cf456b38da64edf1aa025660dd160db116ead31/checks/19148866328/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=kjetilkjeka
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_330bb5a1-d87c-441b-bab5-cba65f058bab
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=33bfc92c32a27c83a8def22969717d5c692552c5
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_330bb5a1-d87c-441b-bab5-cba65f058bab
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_330bb5a1-d87c-441b-bab5-cba65f058bab
GITHUB_TRIGGERING_ACTOR=kjetilkjeka
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/117458/merge
GITHUB_WORKFLOW_SHA=33bfc92c32a27c83a8def22969717d5c692552c5
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
   Compiling rustc-main v0.0.0 (/checkout/compiler/rustc)
    Checking dissimilar v1.0.6
    Checking expect-test v1.4.1
   Compiling coverage_test_macros v0.0.0 (/checkout/compiler/rustc_mir_transform/src/coverage/test_macros)
error[E0599]: no variant or associated item named `Embedded` found for enum `spec::LinkerFlavor` in the current scope
   --> compiler/rustc_target/src/spec/tests/tests_impl.rs:31:87
    |
31  |         assert_eq!(self.arch == "nvptx64", matches!(self.linker_flavor, LinkerFlavor::Embedded));
    |                                                                                       ^^^^^^^^ variant or associated item not found in `LinkerFlavor`
   ::: compiler/rustc_target/src/spec/mod.rs:101:1
    |
101 | pub enum LinkerFlavor {
101 | pub enum LinkerFlavor {
    | --------------------- variant or associated item `Embedded` not found for this enum
For more information about this error, try `rustc --explain E0599`.
error: could not compile `rustc_target` (lib test) due to previous error
warning: build failed, waiting for other jobs to finish...
Build completed unsuccessfully in 0:01:15
```

</details>

kjetilkjeka commented:
@rustbot label -S-waiting-on-author +S-waiting-on-review

Commit:
NVPTX: Enable self-contained for the nvptx target

petrochenkov commented:
>Should I instead also include the distribution artifact in this PR or postpone it to a follow up PR?

Probably postpone and pass it through the infra team, I cannot answer the last question.

bors commented:
:umbrella: The latest upstream changes (presumably #116278) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

juntyr commented:
I tried out the embedded linker crate as a replacement for the old `rust-ptx-linker` on some kernels I've been working on, and it works great! Thank you so much for putting in the effort to get Rust->PTX compilation working smoothly again!

One change I noticed is that the resulting PTX kernels now include extraneous symbols, which were previously internalised by `rust-ptx-linker`'s custom pass. In particular, several `rust_*` symbols, `__rg_oom`, and `rust_begin_unwind` are all included in the final PTX since they are all defined with extern linkage even though they're not used by the kernel. Filtering them out explicitly when the symbols are collected from `llvm-nm` is one approach to handle them, though I wonder if something akin to the original internalising pass could be added again?

kjetilkjeka commented:
Thanks a lot for testing it @juntyr! It's incredibly helpful and very exciting to get some feedback on this stuff.

> One change I noticed is that the resulting PTX kernels now include extraneous symbols, which were previously internalised by rust-ptx-linker's custom pass. In particular, several rust_* symbols, __rg_oom, and rust_begin_unwind are all included in the final PTX since they are all defined with extern linkage even though they're not used by the kernel. Filtering them out explicitly when the symbols are collected from llvm-nm is one approach to handle them, though I wonder if something akin to the original internalising pass could be added again?

The reason the `embedded-linker` uses nm to get a list of external symbols and then uses the `internalize-public-api-file` to perform internalizing is due to this being the way that is supported when calling the llvm binaries. 

The old `rust-ptx-linker` added a custom pass manager to perform internalizing (after some logic is used to mark specific symbols as internal). The ability to define a pass manager to freely combine passes is something that has since been removed from the ("more-stable") llvm C api to reduce the maintenance burden. This is actually the main reason the `rust-ptx-linker` was broken and is no longer usable. The function `LLVMPassManagerBuilderCreate` is no longer available in the C api.

It is definitely possible to re-implement it using the C++ "less-stable" API. This tool currently only calls the llvm tools to keep the maintenance burden very low. I think most things (this included) that is useful to do when generating ptx from Rust will also be useful when generating ptx from other languages and should fit well inside LLVM instead of this tool. I therefore think looking at ways to achieve the same by using functionality in the llvm tools (and if needed, contributing it to the LLVM tools) would be a better fit. I will also admit that I might very well be proven wrong in this assumption

I see that there was a PR to LLVM adding a pass for something like this in AMDGPU. Perhaps adding something similar for NVPTX is a good approach here? I assume this can be added to an invocation of `opt` or `llc`? https://reviews.llvm.org/D29214

In the short term, I think you suggestion of matching the symbols to a rule like `rust_*` or a list of known symbols like `[__rg_oom, _]` and filter them out if a match occurs makes sense.  I think it should be done in a follow up PR.

juntyr commented:
I've experimented a bit more with the embedded linker in a copy of the source code in this PR, which includes the aforementioned symbol exclusion (https://github.com/juntyr/rust-ptx-linker/blob/01aac09d9cfea39976befe1d90d440c28c5614b6/src/embedded_linker/linker.rs#L149-L155).

I also looked into whether I could force LLVM to inline *all* functions (where this is possible, e.g. recursive functions usually break this) by adding the `alwaysinline` attribute to all functions. While this is certainly a quite harsh approach that only makes sense in a few cases, I thought I'd share my diff in the hope that this could perhaps become part of the embedded linker, perhaps under a `--force-inline-all` option:
https://github.com/juntyr/rust-ptx-linker/commit/01aac09d9cfea39976befe1d90d440c28c5614b6

The LLVM passes I am adding after `always-inline` (`gvn`, `globalopt`, `dse`, `globalopt`) are pure guesswork and seem to produce good results for me, but somebody with knowledge of LLVM could definitely come up with something better and more justified.

kjetilkjeka commented:
It's fantastic that you have already started implementing improvements on top of this PR. I'm very happy I'm not the only one that want to see this in the best state it can be. I think enabling the CI to avoid any more stuff breaking on the nvptx target is the highest priority, and I therefore think it makes sense to see this one land in its basic form before any more additions are made. But as soon as this one lands you should open PRs for your additions. I have also started looking at adding it as a component to be able to distribute it to users that I will be opening a PR for when this base implementation lands.

> I've experimented a bit more with the embedded linker in a copy of the source code in this PR, which includes the aforementioned symbol exclusion (https://github.com/juntyr/rust-ptx-linker/blob/01aac09d9cfea39976befe1d90d440c28c5614b6/src/embedded_linker/linker.rs#L149-L155).

Seems sensible to me. I would consider doing it [here](https://github.com/juntyr/rust-ptx-linker/blame/01aac09d9cfea39976befe1d90d440c28c5614b6/src/embedded_linker/linker.rs#L229) since it's a special case when internalizing, but I see the argument for your way as well.

We should probably add some test cases of linking actual code and verifying that the symbols are removed/kept correctly.

> I also looked into whether I could force LLVM to inline all functions (where this is possible, e.g. recursive functions usually break this) by adding the alwaysinline attribute to all functions. While this is certainly a quite harsh approach that only makes sense in a few cases, I thought I'd share my diff in the hope that this could perhaps become part of the embedded linker, perhaps under a --force-inline-all option:
https://github.com/juntyr/rust-ptx-linker/commit/01aac09d9cfea39976befe1d90d440c28c5614b6

I see how forcing the compiler to more aggressively inline functions can be valuable for targets like nvptx. In our internal codebase we have `#[inline(always)]` all over the place and it completely breaks performance if one of these inline attributes are forgotten when writing new code. It's not a great situation having to do this manually.

In your tests, is the `-O3` argument also provided to embedded-linker? I would expect that to perform aggressive inlining on the linked artifact. Especially considering the target is nvptx. But I'm not very familiar with this so my understanding might very well be limited.

Also, I'm uncertain about enabling it on nvptx64 regardless of the `inline` flag since it would work without as well.  

> The LLVM passes I am adding after always-inline (gvn, globalopt, dse, globalopt) are pure guesswork and seem to produce good results for me, but somebody with knowledge of LLVM could definitely come up with something better and more justified.

I'm again interested in seeing what kind of passes using `-O3` for opimization already adds. Other than that I'm afraid I don't have much knowledge of LLVM optimization passes.

juntyr commented:
> It's fantastic that you have already started implementing improvements on top of this PR. I'm very happy I'm not the only one that want to see this in the best state it can be. I think enabling the CI to avoid any more stuff breaking on the nvptx target is the highest priority, and I therefore think it makes sense to see this one land in its basic form before any more additions are made. But as soon as this one lands you should open PRs for your additions. I have also started looking at adding it as a component to be able to distribute it to users that I will be opening a PR for when this base implementation lands.

Sounds great, I'm very much looking forward to seeing this PR land and then later having the embedded linker made available to end users like myself - thank you so much for working on it!

Once the PR has landed, I'll gladly contribute a PR with a more refined version of my ad-hoc improvements.

> I see how forcing the compiler to more aggressively inline functions can be valuable for targets like nvptx. In our internal codebase we have `#[inline(always)]` all over the place and it completely breaks performance if one of these inline attributes are forgotten when writing new code. It's not a great situation having to do this manually.

I am in absolutely the same situation and have kind of grown tired of manually looking into the PTX after every code change or new Rust or LLVM version to see if all functions are still inlined. While even the automatic addition of `alwaysinline` attributes to all functions cannot inline *everything* (e.g. recursive functions), it covers 99% of important cases where inlining failing would otherwise be unexpected.

> In your tests, is the `-O3` argument also provided to embedded-linker? I would expect that to perform aggressive inlining on the linked artifact. Especially considering the target is nvptx. But I'm not very familiar with this so my understanding might very well be limited.

Yes, I'm passing `-O3` in my own tests and while I haven't checked what the PTX would look like without it, it is not sufficient to inline all functions. Some functions are annotated as cold or extern and it requires interning and forceful inlining and at least two rounds of dead code and constant elimination to fully clean up the PTX into a single function kernel.

In the future PR, we should probably add some codegen tests to check that this option would continue working.

bors commented:
:umbrella: The latest upstream changes (presumably #119190) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

petrochenkov commented:
(I should get to this on new year holidays.)

OwenTrokeBillard commented:
Is this and https://github.com/rust-lang/rust/pull/117671 all that's needed to enable the PTX tests again?

kjetilkjeka commented:
This PR actually enables most ptx tests and is therefore all that's needed by itself.

There have unfortunately been some regressions in the ABI part that needs to be looked at in follow ups.

bors commented:
:umbrella: The latest upstream changes (presumably #119556) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

kjetilkjeka commented:
@petrochenkov I guess the new year has been more busy than anticipated. Do you have an estimate of how far you are from looking at this one?

petrochenkov commented:
Not far, hopefully.

In the meantime, I have one question.
How specific the logic in this PR is to nvptx target?
Does providing a rustc interface to linking LLVM bitcode files make sense in general, for other targets?

kjetilkjeka commented:
> In the meantime, I have one question.
> How specific the logic in this PR is to nvptx target?
> Does providing a rustc interface to linking LLVM bitcode files make sense in general, for other targets?

Most of the embedded-linker is not specific to nvptx at all. It is probably specific to embedded (`#[no_std]`) targets.

I have a mental note of making it work with RISC-V targets as well. That would remove the dependency for gcc RISC-V linkers when building code for this target. I have not investigated this beyond the "thinking about it while commuting" stage.

Right now, there are some target dependent "hacks" that are used to workaround deficiencies in the nvptx target. They are automatically applied for `--target nvptx` in the embedded linker and should be removed in the future.

In addition to the general flags there also needs to be target specific flags to adjust target specific behavior. This is already the case in `llc` and friends (e.g. `--nvptx-sched4reg`) which is why I don't see that as a big issue for the embedded-linker either.

petrochenkov commented:
I initially assumed that "embedded" in the name meant that the new linker tool is embedded into the distributed Rust toolchain (sort of similar to "link-self-contained").
But it looks like it actually means that the tool is supposed to be used for embedded targets.
Maybe it's better to rebrand it to something similar to "LLVM bitcode linker", because that better describes what it does?

petrochenkov commented on compiler/rustc_codegen_ssa/src/back/linker.rs:
```diff
@@ -1815,7 +1816,7 @@ impl<'a> Linker for PtxLinker<'a> {
 
     fn link_whole_rlib(&mut self, path: &Path) {
-        self.cmd.arg("--rlib").arg(path);
+        self.cmd.arg("--whole-rlib").arg(path);
```
>I'll try to send a PR adjusting API naming in this file today.

Done in https://github.com/rust-lang/rust/pull/120099.

kjetilkjeka replied:
In a different comment we discuss using a new linker-flavor. If that is the case we can also remove `--rlib` if there's a better way to convey the same information.

When that is said, it's quite convenient that the `llvm-bitcode-linker` knows if it's linking a `rlib` since it is not an entirely standard object archive (it includes some metadata files).

When it comes to `--whole-rlib` an even more important distinction is provided to the linker. It lets it know if the symbols needs to be kept or not. Basically symbols in `core` is not part of the `API` when compiling to a cdylib so these symbols can be internalized while symbols from "user code" needs to be kept. What is the preferred difference in syntax for letting the linker know about this?

kjetilkjeka commented on compiler/rustc_codegen_ssa/src/back/linker.rs:
```diff
@@ -1831,12 +1832,13 @@ impl<'a> Linker for PtxLinker<'a> {
+            OptLevel::Default => "-O2",
+            OptLevel::Aggressive => "-O3",
+            OptLevel::Size => "-Os",
+            OptLevel::SizeMin => "-Oz",
```
I'm very much in favor of adding the new linker flavor from the start.

With a new linker flavor we don't need to worry about breakage between rust-ptx-linker and the new `llvm-bitcode-linker`. When additional target reach some sort of support they can be used with the new linker using `-C linker-flavor llbc` instead of the misleading `-C linker-flavor ptx`. It should be possible to deprecate the ptx-linker flavor quite quickly, but I don't know what's actually required to get rid of it for good.

I see that there is a concept of unstable linker flavor. Since the goal is to distribute `llvm-bitcode-linker` versioned together with rust it would be beneficial to mark it as unstable and avoid making it default for any target yet. This means only (as default) using it for ptx tests at this point, but start recommending it for experimentation for the nvptx target as soon as it can be distributed to users.

kjetilkjeka commented:
Thanks for the review @petrochenkov 

I have added some further questions on your comments related to things that are still unclear to me. These are the main blockers that needs to be resolved for me to progress.

### 1. More specific ptx or more generic
Even though initially only supports `nvptx`, the linker was originally created to be quite generic between a multitude of targets. After it was requested to not introduce more linker-flavors it was adapted to fit the `ptx` linker flavor. I agree with you in your last comments that this semi-specific/semi-generic state is not ideal.

In your comments you basically suggests two directions. Either making everything with the linker very specific to ptx. This consists of renaming it to `rust-ptx-linker`. 

The problem with making it specific is that we need to differentiate between "libs" and "whole libs" to correctly internalize the corelib and must therefore slightly change the linker api. It's also beneficial to have more fine grained control of optimization levels, introducing yet another slight breakage.

The problem with creating a new linker flavor is that then there's yet another linker flavor. Right now my theory of this linker being useful for many targets is basically an unproven hypothesis. If, by making this linker flavor unstable and not a default for any target, it is possible to remove it in the future, it would be less of an disadvantage.

I think there needs to be a decision here before it makes sense to adapt the linker more.  

### 2. `--whole-rlib` alternative syntax
Right now `--whole-rlib` is being passed as an argument. That is the method that is used for figuring out which libraries the symbols can be internalized for.

If `--rlib`/`--whole-rlib` is not satisfactory syntax for this, a different syntax is required. I see that there are some different alternatives in use for different linkers in the wild. I would prefer something that is an entirely different argument instead of something that changes the "state" for all upcoming file (like the wasm linker). But everything is obviously workable.

Do you have any example of acceptable syntax for linking "whole archives"?

kjetilkjeka commented on src/tools/embedded-linker/LICENSE-MIT:
```diff
@@ -0,0 +1,23 @@
+CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR
+IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+DEALINGS IN THE SOFTWARE.
```
I agree that it seems like it's not necessary. I will remove them.

bors commented:
:umbrella: The latest upstream changes (presumably #120341) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

petrochenkov commented:
>1. Make the linker more specific to ptx or more generic?

Let's make it more generic from the start, with a new unstable linker flavor.

petrochenkov commented:
>2. --whole-rlib alternative syntax

On a closer inspection, what this PR does with symbol internalization looks nothing like `--whole-archive`.
With `--rlib` all symbols are internalized by `opt`, but with `--whole-rlib` external symbols are not internalized.

I've seen some discussion about symbol internalization above in the thread, I'll check it a bit later to understand why it's necessary.
In any case, it doesn't look similar to what `--whole-archive` does in other linkers.

petrochenkov commented:
The second issue with `link_whole_rlib` (aka `link_staticlib_by_path(_, true)` after the refactoring) is that it's never actually called for rlibs.
If this functionality is needed for rlibs, then it should be made user-visible somehow.
(I still haven't reread the thread, so not sure which specific rlibs require it.)

kjetilkjeka commented:
The reason why it's especially important to internalize non public symbols for ptx is that you will generally load the module into the driver at startup. This loading consists of heavy processing. Even though lazy loading is just being supported, eager loading is still the standard and a bloated file will add to initialization time.

The reason why it is so important in this "first version" is that it's currently impossible to build the whole corelib for `nvptx64-nvidia-cuda`. At least `core::sync::atomic` is problematic right now. If the corelib is internalized in the linker (before bitcode->ptx compilation) then this bug doesn't surface for programs not containing the unsupported parts of corelib. The bug will needs to be fixed, and this is on my agenda, but re-enabling the unit tests is the highest priority to avoid further breakage.

I tried to do some more research how on `cdylib` are supposed to work. I see from [the original RFC on cdylib](https://rust-lang.github.io/rfcs/1510-cdylib.html) that only symbols marked with `#[no_mangle]` needs to be exported and that "the compiler will also be at liberty to pass extra flags to the linker to actively hide exported Rust symbols from linked libraries". It sounds like having infrastructure in the compiler to tell the linker which symbols is tagged with `#[no_mangle]` would solve the issue generally for cdylibs. Even though ptx especially cares about small libraries, I assume this is something that is cared about more generally. Is this something that exist in some form? Or something that would be beneficial to add?

bjorn3 commented:
The `export_symbols` method on `Linker` is called with the list of symbols that should be exported. Every other symbol should be hidden.

kjetilkjeka commented:
Thanks. That actually makes very much sense now that you said it and I'm looking it at again. I must have been confused by the ways the old linker handled internalizing symbols to not realize initially.

 That means I now have everything I need to make progress.

Commit:
LLVM Bitcode Linker: Added crate

Commit:
LLVM Bitcode Linker: Add as a linker known to the compiler

Commit:
Bootstrap: Add argument for building llvm bitcode linker

kjetilkjeka commented:
I have almost updated the branch with the requested changes. The only thing left is figuring out how to support different linker flavors as we want to use `Ptx` for "normal linking" and `llvm-bitcode-linker` for `-C link-self-contained` type linking.

If there are any examples of a target supporting multiple linker flavors it would be nice to see that. Also general advice is welcome. If not, I will try to read some more code and figure it out tomorrow.

petrochenkov commented:
>use `Ptx` for "normal linking" and `llvm-bitcode-linker` for `-C link-self-contained` type linking.

I wouldn't do this personally, we should either require both `-Clinker-flavor` and `-Clink-self-contained` explicitly, or make `-Clinker-flavor=llvm-bitcode-linker` and `-Clink-self-contained=+linker` the default.

Also, for other targets the linker lookup logic search self-contained directories first, and then search system directories as well if nothing was found. So if the nvptx target follows this approach, then `ptx-linker` will be found even with `-Clink-self-contained=true`.

kjetilkjeka commented:
The problem is that using something like `-Z unstable-options -C linker=llvm-bitcode-linker -C linker-flavor=llvm-bc` gives 

```
error: linker flavor `llvm-bc` is incompatible with the current target
  |
  = note: compatible flavors are: ptx
```

The problem is that everything is engineered around the variants of `LinkerFlavor` representing the target rather than the linker. Then there are members of these variants that define if it is `Lld` or not.

The solution that I think is most intuitive is to change the way we infer and verify the `LinkerFlavor` from `LinkerFlavorCli`. Right now if I use `-C linker-flavor=llvm-bc` I will get `LinkerFlavor::Ptx`  and `LinkerFlavorCli::LlvmBc`. Would it be fine to change it so the `LinkerFlavor` represented the `LinkerFlavorCli`?

Alternatively, the solution could be to change the ptx linker flavor to something like

```rust
#[derive(Clone, Copy, Debug, Eq, Ord, PartialEq, PartialOrd)]
pub enum LinkerFlavor {
    /* Left out variants */
    Ptx(PtxBitcodeLinker),
    /*left out variants */    
}

pub enum PtxBitcodeLinker {
    PtxLinker,
    LlvmBitcodeLinker,
}
```

Or something like 

```rust
#[derive(Clone, Copy, Debug, Eq, Ord, PartialEq, PartialOrd)]
pub enum LinkerFlavor {
    /* Left out variants */
    Bitcode(BitcodeLinker),
    /*left out variants */    
}

pub enum BitcodeLinker {
    PtxLinker,
    BpfLinker,
    LlvmBitcodeLinker,
}
```

Or should perhaps `TargetOptions` be able to list more than one linker?

I think it's worth figuring out now as we are aiming to support more targets with the `llvm-bitcode-linker` without breaking support for the original linkers.

petrochenkov commented:
>The problem is that everything is engineered around the variants of `LinkerFlavor` representing the target rather than the linker.

Yeah, this needs to change, there are other use cases for supporting linkers with "principal flavors" different from the default one for the target, e.g. https://github.com/rust-lang/rust/pull/110869.

I think for now it can just be an exception in `fn check_compatibility` (and other related machinery if necessary).

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7818552156/job/21329041696) [(plain)](https://github.com/rust-lang/rust/commit/80a907038ca8271f5aa4a3384240d7100c67d8e6/checks/21329041696/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=kjetilkjeka
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_cbf1bf66-21ab-4da3-b72c-0b0e85a3ee06
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=bdb57cb29ed68082a0ad5aab8010d986b522061a
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_cbf1bf66-21ab-4da3-b72c-0b0e85a3ee06
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_cbf1bf66-21ab-4da3-b72c-0b0e85a3ee06
GITHUB_TRIGGERING_ACTOR=kjetilkjeka
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/117458/merge
GITHUB_WORKFLOW_SHA=bdb57cb29ed68082a0ad5aab8010d986b522061a
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
#13 3.804 Building wheels for collected packages: reuse
#13 3.805   Building wheel for reuse (pyproject.toml): started
#13 4.136   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 4.137   Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=181117 sha256=f5f58750481f69515c2c0d1d503daf565e2565c370d07fc6aeb95fe3498b4269
#13 4.137   Stored in directory: /tmp/pip-ephem-wheel-cache-dfwwzghe/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
#13 4.139 Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
#13 4.161   Attempting uninstall: setuptools
#13 4.162     Found existing installation: setuptools 59.6.0
#13 4.163     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
---
   Compiling rustc-main v0.0.0 (/checkout/compiler/rustc)
    Checking rustc_builtin_macros v0.0.0 (/checkout/compiler/rustc_builtin_macros)
    Checking dissimilar v1.0.6
    Checking expect-test v1.4.1
error[E0004]: non-exhaustive patterns: `spec::LinkerFlavor::LlvmBc` not covered
   --> compiler/rustc_target/src/spec/tests/tests_impl.rs:43:23
43  |                 match self.linker_flavor {
43  |                 match self.linker_flavor {
    |                       ^^^^^^^^^^^^^^^^^^ pattern `spec::LinkerFlavor::LlvmBc` not covered
note: `spec::LinkerFlavor` defined here
   --> compiler/rustc_target/src/spec/mod.rs:100:10
    |
100 | pub enum LinkerFlavor {
100 | pub enum LinkerFlavor {
    |          ^^^^^^^^^^^^
...
127 |     LlvmBc,
    |     ------ not covered
    = note: the matched value is of type `spec::LinkerFlavor`
help: ensure that all possible cases are being handled by adding a match arm with a wildcard pattern or an explicit pattern as shown
61  ~                     },
61  ~                     },
62  +                     spec::LinkerFlavor::LlvmBc => todo!()

For more information about this error, try `rustc --explain E0004`.
error: could not compile `rustc_target` (lib test) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
```

</details>

kjetilkjeka commented:
@petrochenkov I think all your comments have been addressed now

@juntyr I think some of the decisions around internalizing is changed for the better. You might be interested in seeing if it works closer to how we eventually want it to work now.

@rustbot label -S-waiting-on-author +S-waiting-on-review

kjetilkjeka commented:
@petrochenkov do you know when you will have the chance to take a look on this one again?

petrochenkov commented:
This week, most likely.

petrochenkov commented on compiler/rustc_target/src/spec/mod.rs:
```diff
@@ -480,6 +494,7 @@ linker_flavor_cli_impls! {
     (LinkerFlavorCli::Msvc(Lld::No)) "msvc"
     (LinkerFlavorCli::EmCc) "em-cc"
     (LinkerFlavorCli::Bpf) "bpf"
+    (LinkerFlavorCli::Llbc) "llbc"
```
It's unstable, so it's fine either way.

petrochenkov commented on compiler/rustc_target/src/spec/targets/nvptx64_nvidia_cuda.rs:
```diff
@@ -46,6 +47,9 @@ pub fn target() -> Target {
             supports_stack_protector: false,
 
+            // Support using `self-contained` linkers like the llvm-bitcode-linker
+            link_self_contained: LinkSelfContainedDefault::True,
```
`-Clink-self-contained=+linker` is supposed to ensure that the directory inside Rust distribution in which `llvm-bitcode-linker` binary resides is in `PATH` when rustc is searching for the linker by name.
`-Clink-self-contained=-linker`, on the other hand, is supposed to ensure that the directory inside Rust distribution in which `llvm-bitcode-linker` binary resides is NOT in `PATH` when rustc is searching for the linker by name.

Does it actually work like that now?
I have some suspicion that the directory is always in `PATH`.

kjetilkjeka replied:
It looks you are completely correct. The test still passes with `-Clink-self-contained=-linker`. Do you know if this is a problem with how the bootstrap is implemented for `llvm-bitcode-linker` or the `self-contained` machinery is a whole?

petrochenkov replied:
What is the location of `llvm-bitcode-linker` in the distribution?

All shipped linkers should eventually be placed into the `.../self-contained/linker` subdirectory, which will be added/removed to/from `PATH`, but that's not the case right now because `-Clink-self-contained=+/-linker` is not fully implemented.

petrochenkov replied:
The current situation is probably fine as is for this PR.
But it does make this target spec flag and `-Clink-self-contained` options in tests unnecessary.

kjetilkjeka replied:
> What is the location of llvm-bitcode-linker in the distribution?

Right now it will be put in `bin`.

 Is the correct `self-contained` folder actually `lib/rustlib/<HOST>/lib/self-contained` or is that a different `self-contained` dir?
 
 > But it does make this target spec flag and -Clink-self-contained options in tests unnecessary.
 
I interpret that as you want me to remove them? I guess it will be difficult to make sure they're there for new tests as well so we will need to fix them when this `+linker`/`-linker` works anyway.

kjetilkjeka replied:
I removed the  `-Clink-self-contained` statements now

petrochenkov replied:
> Is the correct `self-contained` folder actually `lib/rustlib/<HOST>/lib/self-contained` or is that a different `self-contained` dir?

I see two self-contained directories in my toolchain - `lib/rustlib/<target>/bin/self-contained` and `lib/rustlib/<target>/lib/self-contained`.
Linkers are supposed to be in the `bin` one, including LLD.

>Is `lld` also supposed to work like this or will `llvm-bitcode-linker` be the first linker to actually work like this? 

Some copies of lld can be found in `lib/rustlib/<target>/bin/gcc-ld` subdirectory, and they indeed work like this, at least when LLD is called through a C/C++ compiler - the `gcc-ld` subdirectory is either considered or not depending on `-Clink-self-contained=+/-linker`.
LLD will also need to be moved to `lib/rustlib/<target>/bin/self-contained/linker` eventually.

kjetilkjeka commented on src/tools/llvm-bitcode-linker/src/linker.rs:
```diff
@@ -0,0 +1,240 @@
+    fn link(&mut self) -> anyhow::Result<()> {
+        tracing::info!("Linking {} bitcode files using llvm-link", self.bitcode.len());
+
+        let llvm_link_output = std::process::Command::new("llvm-link")
```
This was an artifact from when the wrong scheme for internalization was used. I'm fixing this now :+1:

kjetilkjeka commented on src/tools/llvm-bitcode-linker/src/bin/llvm-bitcode-linker.rs:
```diff
@@ -0,0 +1,69 @@
+
+    /// Input Rust rlib archives
+    #[arg(long)]
+    rlib: Vec<PathBuf>,
```
With the correct implementation of internalization it seems like it doesn't even matter if it is an archive or bitcode file. I changed this to simply be `files`

kjetilkjeka commented on src/tools/llvm-bitcode-linker/src/linker.rs:
```diff
@@ -0,0 +1,240 @@
+    pub fn add_bitcode(
+        &mut self,
+        path: impl AsRef<Path>,
+        keep_symbols: bool,
```
Done. I kept the function to avoid having to expose the `Vec<_>` of filepaths directly. On the other side I merged it with the rlib method so one method was removed. Hope this was a fine solution.

kjetilkjeka commented on src/tools/llvm-bitcode-linker/src/linker.rs:
```diff
@@ -0,0 +1,240 @@
+            output_file_link.display(),
+        );
+
+        let link_output = std::process::Command::new("llvm-link")
```
This one was not needed anymore with the other changes.

kjetilkjeka commented on src/tools/llvm-bitcode-linker/src/linker.rs:
```diff
@@ -0,0 +1,240 @@
+    }
+
+    /// Add a single symbol to the list of exported symbols
+    pub fn add_exported_symbol(&mut self, symbol: String) {
```
Removed

Commit:
NVPTX: Enable previously disabled tests

kjetilkjeka commented:
Thanks for the review @petrochenkov 

I made an effort of addressing all your concerns. I have commented on and left open those I were unsure about. But hopefully everything is fine.

Let me know if there's anything more you would like to see in the PR description.

bors commented:
:umbrella: The latest upstream changes (presumably #120468) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

kjetilkjeka commented on src/tools/llvm-bitcode-linker/src/linker.rs:
```diff
@@ -0,0 +1,176 @@
+    fn optimize(
+        &mut self,
+        optimization: Optimization,
+        mut internalize: bool,
```
Removed as the internalize parameter is no longer needed after the explicit export symbol parameter is included.

kjetilkjeka commented on src/tools/llvm-bitcode-linker/src/linker.rs:
```diff
@@ -0,0 +1,176 @@
+        self.symbols.extend(symbols);
+    }
+
+    fn link(&mut self) -> anyhow::Result<()> {
```
Done

kjetilkjeka commented:
@petrochenkov I've made another attempt to address the remaining comments. Hope this is fine. 

I'm also working on making the `+linker/-linker` stuff work, but plan to add this in a new PR as previously discussed

@rustbot label -S-waiting-on-author +S-waiting-on-review

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-16`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8172079413/job/22341667399) [(plain)](https://github.com/rust-lang/rust/commit/248bb2b66f2fecf68b39908950317011f3259262/checks/22341667399/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=kjetilkjeka
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_2709c073-7c26-45d3-8347-4ee2c3c7b8fe
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=95e76257c6de82a5560b98d33c7a99e68c02f255
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_2709c073-7c26-45d3-8347-4ee2c3c7b8fe
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_2709c073-7c26-45d3-8347-4ee2c3c7b8fe
GITHUB_TRIGGERING_ACTOR=kjetilkjeka
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/117458/merge
GITHUB_WORKFLOW_SHA=95e76257c6de82a5560b98d33c7a99e68c02f255
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_20_X64=/opt/hostedtoolcache/go/1.20.14/x64
---
#12 writing image sha256:1a73fc264ddfce6a204701b960584e3e34e0778452968d5ad3ac5f624ce5264d done
#12 naming to docker.io/library/rust-ci done
#12 DONE 10.2s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-16]
##[group]Clock drift check
  local time: Wed Mar  6 12:32:12 UTC 2024
  network time: Wed, 06 Mar 2024 12:32:12 GMT
  network time: Wed, 06 Mar 2024 12:32:12 GMT
##[endgroup]
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-16', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'build.optimized-compiler-builtins', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-16/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
For more information about this error, try `rustc --explain E0609`.
error: could not compile `llvm-bitcode-linker` (bin "llvm-bitcode-linker") due to 1 previous error


command did not execute successfully: cd "/checkout" && env -u MAKEFLAGS -u MFLAGS AR_x86_64_unknown_linux_gnu="ar" CARGO_INCREMENTAL="0" CARGO_PROFILE_RELEASE_DEBUG="0" CARGO_PROFILE_RELEASE_DEBUG_ASSERTIONS="true" CARGO_PROFILE_RELEASE_OVERFLOW_CHECKS="true" CARGO_PROFILE_RELEASE_STRIP="false" CARGO_TARGET_DIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools" CC_x86_64_unknown_linux_gnu="sccache cc" CFG_COMPILER_BUILD_TRIPLE="x86_64-unknown-linux-gnu" CFG_COMPILER_HOST_TRIPLE="x86_64-unknown-linux-gnu" CFG_RELEASE="1.78.0-nightly" CFG_RELEASE_CHANNEL="nightly" CFG_RELEASE_NUM="1.78.0" CFG_VERSION="1.78.0-nightly (95e76257c 2024-03-06)" CFG_VER_DATE="2024-03-06" CFG_VER_HASH="95e76257c6de82a5560b98d33c7a99e68c02f255" CFLAGS_x86_64_unknown_linux_gnu="-ffunction-sections -fdata-sections -fPIC -m64" CXXFLAGS_x86_64_unknown_linux_gnu="-ffunction-sections -fdata-sections -fPIC -m64" CXX_x86_64_unknown_linux_gnu="sccache c++" DOC_RUST_LANG_ORG_CHANNEL="https://doc.rust-lang.org/nightly" LIBC_CHECK_CFG="1" LIBRARY_PATH="/usr/lib/llvm-16/lib" LZMA_API_STATIC="1" RANLIB_x86_64_unknown_linux_gnu="ar s" REAL_LIBRARY_PATH_VAR="LD_LIBRARY_PATH" RUSTBUILD_NATIVE_DIR="/checkout/obj/build/x86_64-unknown-linux-gnu/native" RUSTC="/checkout/obj/build/bootstrap/debug/rustc" RUSTC_BOOTSTRAP="1" RUSTC_BREAK_ON_ICE="1" RUSTC_ERROR_METADATA_DST="/checkout/obj/build/tmp/extended-error-metadata" RUSTC_HOST_FLAGS="--cfg=bootstrap -Zunstable-options --check-cfg=cfg(bootstrap)" RUSTC_INSTALL_BINDIR="bin" RUSTC_LIBDIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" RUSTC_LINT_FLAGS="-Wrust_2018_idioms -Wunused_lifetimes -Dwarnings" RUSTC_REAL="/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustc" RUSTC_SNAPSHOT="/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustc" RUSTC_SNAPSHOT_LIBDIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" RUSTC_STAGE="0" RUSTC_SYSROOT="/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-sysroot" RUSTC_TLS_MODEL_INITIAL_EXEC="1" RUSTC_VERBOSE="0" RUSTC_WRAPPER="/checkout/obj/build/bootstrap/debug/rustc" RUSTDOC="/checkout/obj/build/bootstrap/debug/rustdoc" RUSTDOCFLAGS="--cfg=bootstrap --cfg=windows_raw_dylib -Csymbol-mangling-version=v0 -Zunstable-options --check-cfg=cfg(bootstrap) --check-cfg=cfg(parallel_compiler) --check-cfg=cfg(rust_analyzer) --check-cfg=cfg(no_btreemap_remove_entry) --check-cfg=cfg(crossbeam_loom) --check-cfg=cfg(span_locations) --check-cfg=cfg(rustix_use_libc) --check-cfg=cfg(emulate_second_only_system) --check-cfg=cfg(windows_raw_dylib) -Dwarnings -Wrustdoc::invalid_codeblock_attributes --crate-version 1.78.0-nightly\t(95e76257c\t2024-03-06)" RUSTDOC_REAL="/path/to/nowhere/rustdoc/not/required" RUSTFLAGS="--cfg=bootstrap --cfg=windows_raw_dylib -Csymbol-mangling-version=v0 -Zunstable-options --check-cfg=cfg(bootstrap) --check-cfg=cfg(parallel_compiler) --check-cfg=cfg(rust_analyzer) --check-cfg=cfg(no_btreemap_remove_entry) --check-cfg=cfg(crossbeam_loom) --check-cfg=cfg(span_locations) --check-cfg=cfg(rustix_use_libc) --check-cfg=cfg(emulate_second_only_system) --check-cfg=cfg(windows_raw_dylib) -Zmacro-backtrace -Csplit-debuginfo=off -Cllvm-args=-import-instr-limit=10 -Clink-args=-Wl,-z,origin -Clink-args=-Wl,-rpath,$ORIGIN/../lib -Zunstable-options" RUST_TEST_THREADS="16" SYSROOT="/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-sysroot" TERM="xterm" __CARGO_DEFAULT_LIB_METADATA="bootstraptool-rustc" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "build" "--target" "x86_64-unknown-linux-gnu" "--release" "-Zcheck-cfg" "-Zbinary-dep-depinfo" "-j" "16" "--locked" "--color" "always" "--manifest-path" "/checkout/src/tools/llvm-bitcode-linker/Cargo.toml"
Build completed unsuccessfully in 0:03:36

stdout ----

```

</details>

petrochenkov commented:
This should be ready now.
Thanks!
@bors r+

bors commented:
:pushpin: Commit 8b47f5a7fc209b6c289850c30693f60b7772a312 has been approved by `petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov 8b47f5a7fc209b6c289850c30693f60b7772a312 -->
<!-- homu: {"type":"Approved","sha":"8b47f5a7fc209b6c289850c30693f60b7772a312","approver":"petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #122113) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

kjetilkjeka commented:
I did a manual rebase to fix the conflict now. It was only a minor conflict (in `change_trackers.rs`) so this one should still be ready.

@rustbot label -S-waiting-on-author +S-waiting-on-review

petrochenkov commented:
@bors r+

bors commented:
:pushpin: Commit 355e83361e8f1ec452fc4bae7426733c87f81885 has been approved by `petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov 355e83361e8f1ec452fc4bae7426733c87f81885 -->
<!-- homu: {"type":"Approved","sha":"355e83361e8f1ec452fc4bae7426733c87f81885","approver":"petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

kjetilkjeka commented:
@petrochenkov Yet another conflict in `change_trackers.rs` resolved.

@rustbot label -S-waiting-on-author +S-waiting-on-review

petrochenkov commented:
@bors r+

bors commented:
:pushpin: Commit 843dd28d21d410d540515cf44172ad74ab5976c9 has been approved by `petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov 843dd28d21d410d540515cf44172ad74ab5976c9 -->
<!-- homu: {"type":"Approved","sha":"843dd28d21d410d540515cf44172ad74ab5976c9","approver":"petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
